source/types.ts

```diff
@@ -0,0 +1,247 @@
+// /source/types.ts
+// All the types used by this package
+
+import Express from 'express'
+
+/**
+ * Callback that fires when a client's hit counter is incremented.
+ *
+ * @param error {Error | undefined} - The error that occurred, if any
+ * @param hit {number} - The number of hits for that client so far
+ * @param resetTime {Date | undefined} - The time when the counter resets
+ *
+ * @public
+ */
+export type IncrementCallback = (
+	error: Error | undefined,
+	hit: number,
+	resetTime: Date | undefined,
+) => void
+
+/**
+ * Method (in the form of middleware) to generate custom identifiers for
+ * clients.
+ *
+ * @param request {Express.Request} - The Express request object
+ * @param response {Express.Response} - The Express response object
+ *
+ * @returns {string} - The string used to identify the client
+ *
+ * @public
+ */
+export type KeyGeneratingMiddleware = (
+	request: Express.Request,
+	response: Express.Response,
+) => string | Promise<string>
+
+/**
+ * Method (in the form of middleware) to determine whether or not this request
+ * counts towards a client's quota.
+ *
+ * @param request {Express.Request} - The Express request object
+ * @param response {Express.Response} - The Express response object
+ *
+ * @returns {boolean} - Whether or not this request counts towards the quota
+ *
+ * @public
+ */
+export type ShouldSkipMiddleware = (
+	request: Express.Request,
+	response: Express.Response,
+) => boolean | Promise<boolean>
+
+/**
+ * A modified Express request handler with the rate limit functions.
+ */
+export type RateLimitRequestHandler = Express.RequestHandler & {
+	/**
+	 * Method to reset a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 */
+	resetKey: (key: string) => void
+}
+
+/**
+ * An interface that all hit counter stores must implement.
+ */
+export interface Store {
+	/**
+	 * Method to increment a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 * @param callback {IncrementCallback} - The callback to call once the counter is incremented
+	 */
+	increment: (key: string, callback: IncrementCallback) => void
```

Comment:
As I mentioned above, we need to also support legacy stores with the `incr()` method. 

Is there a way of marking that as deprecated, or telling typescript that it needs to have one or the other of `incr`, `increment` (but not both)?  Maybe just make them both option and explain it in the documentation otherwise.

Reply:
We could mark `incr` as deprecated but still support it if it exists. Something like this:

```ts
/**
 * An interface that all hit counter stores must implement.
 */
export interface Store {
	/**
	 * (Deprecated) Method to increment a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client
	 * @param callback {IncrementCallback} - The callback to call once the counter is incremented
	 *
	 * @deprecated 6.x - Implement the `increment` function instead.
	 */
	incr?: (key: string, callback: IncrementCallback) => void

	/**
	 * Method to increment a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client
	 * @param callback {IncrementCallback} - The callback to call once the counter is incremented
	 */
	increment: (key: string, callback: IncrementCallback) => void

	// ...
}
```

And then replace [this](https://github.com/gamemaker1/express-rate-limit/blob/opiniated-improvements/source/lib.ts#L148) with:

```ts
(options.store.increment ?? options.store.incr)(
	...
)
```

Does that sound good?

Reply:
Yea, I'm OK with either this approach or the wrapper. I kind of like the wrapper better, I could live with either one.

Reply:
> Going one step further, we can probably automatically detect and wrap stores using the legacy API. So end-users can upgrade with no code changes.

I went ahead and implemented a basic `promisifyStore` function to wrap up legacy stores. Will add the `store.init` part now.